// akita-filters-plugin/esm5/lib/akita-filters-plugin.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { AkitaFiltersStore, createFilter } from './akita-filters-store';
import { AkitaFiltersQuery } from './akita-filters-query';
import { combineLatest, isObservable, merge } from 'rxjs';
import { map } from 'rxjs/operators';
import { compareValues, EntityCollectionPlugin, isFunction, isUndefined } from '@datorama/akita';
/**
 * @record
 * @template S
 */
export function FiltersParams() { }
if (false) {
    /** @type {?|undefined} */
    FiltersParams.prototype.filtersStoreName;
    /** @type {?|undefined} */
    FiltersParams.prototype.entityIds;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
function NormalizedFilterOptions() { }
if (false) {
    /** @type {?|undefined} */
    NormalizedFilterOptions.prototype.withSort;
    /** @type {?|undefined} */
    NormalizedFilterOptions.prototype.asQueryParams;
    /** @type {?|undefined} */
    NormalizedFilterOptions.prototype.sortByKey;
    /** @type {?|undefined} */
    NormalizedFilterOptions.prototype.sortByOrderKey;
}
/**
 * @template S, E, I, P
 */
var /**
 * @template S, E, I, P
 */
AkitaFiltersPlugin = /** @class */ (function (_super) {
    tslib_1.__extends(AkitaFiltersPlugin, _super);
    function AkitaFiltersPlugin(query, params) {
        if (params === void 0) { params = {}; }
        var _this = _super.call(this, query, params.entityIds) || this;
        _this.query = query;
        _this.params = params;
        _this._server = false;
        _this.params = tslib_1.__assign({ filtersStoreName: _this.getStore().storeName + 'Filters' }, params);
        _this._filtersStore = new AkitaFiltersStore(_this.params.filtersStoreName);
        _this._filtersQuery = new AkitaFiltersQuery(_this._filtersStore);
        _this._selectFilters$ = _this.filtersQuery.selectAll({ sortBy: 'order' });
        _this._selectFiltersAll$ = _this.filtersQuery.selectAll({ sortBy: 'order', filterBy: (/**
             * @param {?} filter
             * @return {?}
             */
            function (filter) { return !filter.hide; }) });
        _this._selectSortBy$ = _this.filtersQuery.select((/**
         * @param {?} state
         * @return {?}
         */
        function (state) { return state && state.sort ? state.sort : null; }));
        return _this;
    }
    Object.defineProperty(AkitaFiltersPlugin.prototype, "filtersStore", {
        get: /**
         * @return {?}
         */
        function () {
            return this._filtersStore;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AkitaFiltersPlugin.prototype, "filtersQuery", {
        get: /**
         * @return {?}
         */
        function () {
            return this._filtersQuery;
        },
        enumerable: true,
        configurable: true
    });
    /**
     *  Add support of filters from server. Provide a function that will be call each time a filter changes
     *
     *  new AkitaFilterPlugins(query).withServer((filters) => {
     *      return this.api.getData(filters);
     *  });
     *
     *  Return false to not add in store. if you want to manage the store in your own.
     */
    /**
     *  Add support of filters from server. Provide a function that will be call each time a filter changes
     *
     *  new AkitaFilterPlugins(query).withServer((filters) => {
     *      return this.api.getData(filters);
     *  });
     *
     *  Return false to not add in store. if you want to manage the store in your own.
     * @param {?} onChangeFilter
     * @param {?=} options
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.withServer = /**
     *  Add support of filters from server. Provide a function that will be call each time a filter changes
     *
     *  new AkitaFilterPlugins(query).withServer((filters) => {
     *      return this.api.getData(filters);
     *  });
     *
     *  Return false to not add in store. if you want to manage the store in your own.
     * @param {?} onChangeFilter
     * @param {?=} options
     * @return {?}
     */
    function (onChangeFilter, options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        this._server = true;
        // Change default select filters to remove server filters, if you use selectAllByFilters();
        this._selectFilters$ = this._filtersQuery.selectAll({ sortBy: 'order', filterBy: (/**
             * @param {?} filter
             * @return {?}
             */
            function (filter) { return !filter.server; }) });
        /** @type {?} */
        var listObservable = [];
        listObservable.push(this._filtersQuery.selectAll({ sortBy: 'order', filterBy: (/**
             * @param {?} filter
             * @return {?}
             */
            function (filter) { return filter.server; }) }));
        if (options.withSort) {
            listObservable.push(this.selectSortBy());
        }
        merge(listObservable).subscribe((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var returnOnChange = onChangeFilter(_this.getNormalizedFilters(options));
            if (returnOnChange !== false && isObservable(returnOnChange)) {
                returnOnChange.subscribe((/**
                 * @param {?} newValue
                 * @return {?}
                 */
                function (newValue) {
                    _this.getStore().set(newValue);
                }));
            }
        }));
        return this;
    };
    /** Return true, if server is configured **/
    /**
     * Return true, if server is configured *
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.hasServer = /**
     * Return true, if server is configured *
     * @return {?}
     */
    function () {
        return this._server;
    };
    /**
     *  Select all filters
     *
     *  Note: Only all filters not hided (with hide=true), will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.selectAll()`
     *
     *
     */
    /**
     *  Select all filters
     *
     *  Note: Only all filters not hided (with hide=true), will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.selectAll()`
     *
     *
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.selectFilters = /**
     *  Select all filters
     *
     *  Note: Only all filters not hided (with hide=true), will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.selectAll()`
     *
     *
     * @return {?}
     */
    function () {
        return this._selectFiltersAll$;
    };
    /**
     * Get all the current snapshot filters
     *
     *  Note: filters with hide=true, will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.getAll()`
     */
    /**
     * Get all the current snapshot filters
     *
     *  Note: filters with hide=true, will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.getAll()`
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.getFilters = /**
     * Get all the current snapshot filters
     *
     *  Note: filters with hide=true, will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.getAll()`
     * @return {?}
     */
    function () {
        return this._filtersQuery.getAll({ filterBy: (/**
             * @param {?} filter
             * @return {?}
             */
            function (filter) { return !filter.hide; }) });
    };
    /**
     * Get all the current snapshot server filters (only if server is available else return default not hidden filters)
     *
     *  Note: filters with server=false, will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.getAll()`
     */
    /**
     * Get all the current snapshot server filters (only if server is available else return default not hidden filters)
     *
     *  Note: filters with server=false, will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.getAll()`
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.getServerFilters = /**
     * Get all the current snapshot server filters (only if server is available else return default not hidden filters)
     *
     *  Note: filters with server=false, will not be displayed. If you want it, call directly to:
     * `this.filtersQuery.getAll()`
     * @return {?}
     */
    function () {
        return this._server ? this._filtersQuery.getAll({ filterBy: (/**
             * @param {?} filter
             * @return {?}
             */
            function (filter) { return !filter.server; }) }) : this.getFilters();
    };
    /**
     * Select All Entity with apply filter to it, and updated with any change (entity or filter)
     * Will not apply sort, if need return   asObject:true !
     */
    /**
     * Select All Entity with apply filter to it, and updated with any change (entity or filter)
     * Will not apply sort, if need return   asObject:true !
     * @param {?=} options
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.selectAllByFilters = /**
     * Select All Entity with apply filter to it, and updated with any change (entity or filter)
     * Will not apply sort, if need return   asObject:true !
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        var _this = this;
        if (options && options.asObject) {
            return combineLatest(this._selectFilters$, this.getQuery().selectAll(options)).pipe(map((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 2), filters = _b[0], entities = _b[1];
                /** @type {?} */
                var unkNowEntity = entities;
                return _this._applyFiltersForHashMap(((/** @type {?} */ (unkNowEntity))), filters);
            })));
        }
        else {
            return combineLatest(this._selectFilters$, this.getQuery().selectAll(options), this.selectSortBy()).pipe(map((/**
             * @param {?} __0
             * @return {?}
             */
            function (_a) {
                var _b = tslib_1.__read(_a, 3), filters = _b[0], entities = _b[1], sort = _b[2];
                /** @type {?} */
                var unkNowEntity = entities;
                return _this._applyFiltersForArray(((/** @type {?} */ (unkNowEntity))), filters, sort);
            })));
        }
    };
    /**
     * Create or update a filter
     */
    /**
     * Create or update a filter
     * @param {?} filter
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.setFilter = /**
     * Create or update a filter
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        if (this._server && isUndefined(typeof filter.server)) {
            filter.server = true;
        }
        /** @type {?} */
        var entity = createFilter(filter);
        this.filtersStore.upsert(entity.id, entity);
    };
    /**
     * Remove a Filter
     */
    /**
     * Remove a Filter
     * @param {?} id
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.removeFilter = /**
     * Remove a Filter
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.filtersStore.remove(id);
    };
    /**
     * Clear all filters
     */
    /**
     * Clear all filters
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.clearFilters = /**
     * Clear all filters
     * @return {?}
     */
    function () {
        this.filtersStore.remove();
    };
    /**
     * Get filter value, return null, if value not available
     */
    /**
     * Get filter value, return null, if value not available
     * @template T
     * @param {?} id
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.getFilterValue = /**
     * Get filter value, return null, if value not available
     * @template T
     * @param {?} id
     * @return {?}
     */
    function (id) {
        if (this.filtersQuery.hasEntity(id)) {
            /** @type {?} */
            var entity = this.filtersQuery.getEntity(id);
            return entity.value ? entity.value : null;
        }
        return null;
    };
    /**
     * Get filter value, return null, if value not available
     */
    /**
     * Get filter value, return null, if value not available
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.getSortValue = /**
     * Get filter value, return null, if value not available
     * @return {?}
     */
    function () {
        /** @type {?} */
        var state = this.filtersQuery.getValue();
        return state.sort ? state.sort : null;
    };
    /**
     * Select Sort by value
     */
    /**
     * Select Sort by value
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.selectSortBy = /**
     * Select Sort by value
     * @return {?}
     */
    function () {
        return this._selectSortBy$;
    };
    /**
     * Set orderBy
     */
    /**
     * Set orderBy
     * @param {?} order
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.setSortBy = /**
     * Set orderBy
     * @param {?} order
     * @return {?}
     */
    function (order) {
        this.filtersStore.update({ sort: order });
    };
    /**
     * Get the filters normalized as key value or as query params.
     * This can be useful for server-side filtering
     */
    /**
     * Get the filters normalized as key value or as query params.
     * This can be useful for server-side filtering
     * @param {?=} options
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.getNormalizedFilters = /**
     * Get the filters normalized as key value or as query params.
     * This can be useful for server-side filtering
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        var e_1, _a;
        if (options === void 0) { options = {}; }
        /** @type {?} */
        var result = {};
        options = tslib_1.__assign({ sortByKey: 'sortBy', sortByOrderKey: 'sortByOrder' }, options);
        try {
            for (var _b = tslib_1.__values(this.getServerFilters()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var filter = _c.value;
                result[filter.id] = filter.value;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (options.withSort) {
            /** @type {?} */
            var sort = this.getSortValue();
            result[options.sortByKey] = sort.sortBy;
            result[options.sortByOrderKey] = sort.sortByOrder;
        }
        if (options.asQueryParams) {
            return this._serialize(result);
        }
        return result;
    };
    /**
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.destroy = /**
     * @return {?}
     */
    function () {
        this.clearFilters();
    };
    /** This method is responsible for getting access to the query. */
    /**
     * This method is responsible for getting access to the query.
     * @protected
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.getQuery = /**
     * This method is responsible for getting access to the query.
     * @protected
     * @return {?}
     */
    function () {
        return this.query;
    };
    /** This method is responsible for getting access to the store. */
    /**
     * This method is responsible for getting access to the store.
     * @protected
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.getStore = /**
     * This method is responsible for getting access to the store.
     * @protected
     * @return {?}
     */
    function () {
        return this.getQuery().__store__;
    };
    /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    AkitaFiltersPlugin.prototype._serialize = /**
     * @private
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        return Object.keys(obj)
            .map((/**
         * @param {?} k
         * @return {?}
         */
        function (k) { return encodeURIComponent(k) + "=" + encodeURIComponent(obj[k]); }))
            .join('&');
    };
    /**
     * @private
     * @param {?} entities
     * @param {?} filters
     * @param {?} sort
     * @return {?}
     */
    AkitaFiltersPlugin.prototype._applyFiltersForArray = /**
     * @private
     * @param {?} entities
     * @param {?} filters
     * @param {?} sort
     * @return {?}
     */
    function (entities, filters, sort) {
        /** @type {?} */
        var entitiesFiltered = entities;
        if (filters.length !== 0) {
            entitiesFiltered = entities.filter((/**
             * @param {?} entity
             * @param {?} index
             * @param {?} array
             * @return {?}
             */
            function (entity, index, array) {
                return filters.every((/**
                 * @param {?} filter
                 * @return {?}
                 */
                function (filter) {
                    if (filter.predicate) {
                        return !!filter.predicate(entity, index, array, filter);
                    }
                    return true;
                }));
            }));
        }
        if (sort && sort.sortBy) {
            /** @type {?} */
            var _sortBy_1 = isFunction(sort.sortBy) ? sort.sortBy : compareValues(sort.sortBy, sort.sortByOrder);
            entitiesFiltered = tslib_1.__spread(entitiesFiltered.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) { return _sortBy_1(a, b, entities); })));
        }
        return entitiesFiltered;
    };
    /**
     * @private
     * @param {?} entities
     * @param {?} filters
     * @return {?}
     */
    AkitaFiltersPlugin.prototype._applyFiltersForHashMap = /**
     * @private
     * @param {?} entities
     * @param {?} filters
     * @return {?}
     */
    function (entities, filters) {
        var _this = this;
        if (filters.length === 0) {
            return entities;
        }
        /** @type {?} */
        var hashMapFiltered = {};
        Object.keys(entities).forEach((/**
         * @param {?} entityKey
         * @param {?} index
         * @return {?}
         */
        function (entityKey, index) {
            /** @type {?} */
            var entity = (/** @type {?} */ (entities[entityKey]));
            if (_this._applyFiltersForOneEntity(filters, entity, index, entities)) {
                hashMapFiltered[entityKey] = entity;
            }
        }));
        return hashMapFiltered;
    };
    /**
     * @private
     * @param {?} filters
     * @param {?} entity
     * @param {?} index
     * @param {?} array
     * @return {?}
     */
    AkitaFiltersPlugin.prototype._applyFiltersForOneEntity = /**
     * @private
     * @param {?} filters
     * @param {?} entity
     * @param {?} index
     * @param {?} array
     * @return {?}
     */
    function (filters, entity, index, array) {
        return filters.every((/**
         * @param {?} filter
         * @return {?}
         */
        function (filter) {
            if (filter.predicate) {
                return !!filter.predicate(entity, index, array, filter);
            }
            return true;
        }));
    };
    /**
     * @protected
     * @param {?} id
     * @return {?}
     */
    AkitaFiltersPlugin.prototype.instantiatePlugin = /**
     * @protected
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return null;
    };
    return AkitaFiltersPlugin;
}(EntityCollectionPlugin));
/**
 * @template S, E, I, P
 */
export { AkitaFiltersPlugin };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AkitaFiltersPlugin.prototype._filtersStore;
    /**
     * @type {?}
     * @private
     */
    AkitaFiltersPlugin.prototype._filtersQuery;
    /**
     * @type {?}
     * @private
     */
    AkitaFiltersPlugin.prototype._server;
    /**
     * @type {?}
     * @private
     */
    AkitaFiltersPlugin.prototype._selectFilters$;
    /**
     * @type {?}
     * @private
     */
    AkitaFiltersPlugin.prototype._selectSortBy$;
    /**
     * @type {?}
     * @private
     */
    AkitaFiltersPlugin.prototype._selectFiltersAll$;
    /**
     * @type {?}
     * @protected
     */
    AkitaFiltersPlugin.prototype.query;
    /**
     * @type {?}
     * @private
     */
    AkitaFiltersPlugin.prototype.params;
}
