// akita-filters-plugin/esm2015/lib/akita-filters-store.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { defaultFilter } from './filters-utils';
import { EntityStore, guid, StoreConfig } from '@datorama/akita';
/**
 * @param {?} str
 * @return {?}
 */
function capitalize(str) {
    return str.charAt(0).toUpperCase() + str.substr(1);
}
/**
 * @record
 * @template S, E
 */
export function AkitaFilter() { }
if (false) {
    /** @type {?} */
    AkitaFilter.prototype.id;
    /**
     * A corresponding name for display the filter, by default, it will be ${id): ${value}
     * @type {?|undefined}
     */
    AkitaFilter.prototype.name;
    /**
     * set the order for filter, by default, it is 10
     * @type {?|undefined}
     */
    AkitaFilter.prototype.order;
    /**
     * The filter value, this will be used to compute name, or getting the current value, to initiate your form
     * @type {?|undefined}
     */
    AkitaFilter.prototype.value;
    /**
     * If you want to have filter that is not displayed on the list
     * @type {?|undefined}
     */
    AkitaFilter.prototype.hide;
    /**
     * If you have enabled server filter, specify witch filters will be call to server, default to false.
     * @type {?|undefined}
     */
    AkitaFilter.prototype.server;
    /**
     * The function to apply filters, by default use defaultFilter helpers, that will search the value in the object
     * @type {?}
     */
    AkitaFilter.prototype.predicate;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @template S, E
 * @param {?} filterParams
 * @return {?}
 */
export function createFilter(filterParams) {
    /** @type {?} */
    const id = filterParams.id ? filterParams.id : guid();
    /** @type {?} */
    const name = filterParams.name || (filterParams.value && filterParams.id ?
        `${capitalize(filterParams.id.toString())}: ${filterParams.value.toString()}` : undefined);
    if (!filterParams.predicate && filterParams.value) {
        /** use default function, if not provided */
        // @ts-ignore
        filterParams.predicate = defaultFilter;
    }
    return (/** @type {?} */ (Object.assign({ id, name, hide: false, order: 10, server: false }, filterParams)));
}
/**
 * @record
 * @template S, E
 */
export function FiltersState() { }
if (false) {
    /** @type {?} */
    FiltersState.prototype.sort;
}
/**
 * @template S
 */
let AkitaFiltersStore = /**
 * @template S
 */
class AkitaFiltersStore extends EntityStore {
    /**
     * @param {?} storeName
     */
    constructor(storeName) {
        super(undefined, { name: storeName });
    }
};
AkitaFiltersStore.ctorParameters = () => [
    { type: String }
];
/**
 * @template S
 */
AkitaFiltersStore = tslib_1.__decorate([
    StoreConfig({ name: 'filters' }),
    tslib_1.__metadata("design:paramtypes", [String])
], AkitaFiltersStore);
export { AkitaFiltersStore };
